### PR TITLE
feat(user): ensure a user document is always present

### LINKED ISSUE #84: feat(user): prevent deleting of users when only one user remains.
Prevent the system getting into a state with no users.

### PR DESCRIPTION
**Pull request by caspiano**
Creates a `before_destroy` callback that checks if at least one `User` remains before destroying the document.
This prevents the system from entering a state with no users.

Resolves #84

### CODE REVIEW

File: spec/user_spec.cr [hunk 1]
```diff
@@ -22,6 +22,23 @@ module PlaceOS::Model
...
+        user = Generator.user.save!
+        expect_raises(Model::Error, "At least one user must remain") do
+          user.destroy
+        end
+      end
+
+      it "does not raise if more than one User remains" do
+        User.clear
+        user0 = Generator.user.save!
+        Generator.user.save!
```
Comment (dukeraphaelng) [inline spec/user_spec.cr:37]: duplicate?

File: spec/authority_spec.cr [hunk 2]
```diff
@@ -49,11 +49,24 @@ module PlaceOS::Model
...
+        it "removes dependent users" do
+          # Keep a dummy user to ensure authority's user can be deleted
+          Generator.user.save!
+          auth = Generator.authority.save!
+          user = Generator.user(authority: auth).save!
+          auth.destroy
+          User.find(user.id.as(String)).should be_nil
+        end
+
+        it "fails if single user remains" do
```
Comment (kimburgess) [inline spec/authority_spec.cr:62]:
```suggestion
        it "fails if no users remain" do
```

File: spec/user_spec.cr (hunk 1 above)
Comment (kimburgess) [reply spec/user_spec.cr:37]: @dukeraphaelng this is generatating two user's, then deleting one.

File: src/placeos-models/user.cr [hunk 3]
```diff
@@ -85,11 +85,17 @@ module PlaceOS::Model
...
 
+    before_destroy :ensure_user_remains
     before_destroy :cleanup_auth_tokens
     before_save :build_name
     before_save :create_email_digest
 
-    # deletes auth tokens
+    # Prevent the system from entering state with no users
+    protected def ensure_user_remains
+      raise Model::Error.new("At least one user must remain") unless User.count > 1
```
Comment (kimburgess) [inline src/placeos-models/user.cr:95]: Do we need to worry about concurrent destroys here?

**Review by kimburgess — CHANGES_REQUESTED** [review]
Worth ammending this to ensure that at least one _admin_ level user remains.

File: src/placeos-models/user.cr (hunk 3 above)
Comment (caspiano) [reply src/placeos-models/user.cr:95]: I have implemented db locks for rethinkORM, so we could use that. Might leave that as a todo for now, and solve it in another PR.

File: src/placeos-models/user.cr [hunk 4]
```diff
@@ -85,11 +86,48 @@ module PlaceOS::Model
...
+
+    # Prevent the system from entering a state with no admin
+    protected def ensure_admin_remains
+      return unless self.sys_admin
+
+      if User.count(sys_admin: true) == 1
+        admin_destroy_lock.unlock if admin_destroy_lock.locked?
+        raise Model::Error.new("At least one admin must remain")
+      end
+    end
```
Comment (kimburgess) [inline src/placeos-models/user.cr:128]: To reduce some of that manual lock management, what about?
```suggestion
    def destroy
      return super unless self.sys_admin
      
      admin_destroy_lock.synchronize do
        if User.count(sys_admin: true) == 1
          raise Model::Error.new("At least one admin must remain")
        end
        super
      end
    end
```

**Review by kimburgess — APPROVED** [review]

File: src/placeos-models/user.cr (hunk 4 above)
Comment (caspiano) [reply src/placeos-models/user.cr:128]: I would prefer to keep the actual check within a callback, however, the synchronisation in the overridden `destroy` is definitely simpler (despite duplicating the sys_admin check)

Comment (caspiano) [reply src/placeos-models/user.cr:128]: Something along the lines of `around_*` callbacks would be useful to prevent patching orm methods.

Comment (kimburgess) [reply src/placeos-models/user.cr:128]: Agreed - had a look for exactly that!

**Pull request merged**